enhancements/microshift/microshift-apiserver-certs.md

```diff
@@ -0,0 +1,302 @@
+certificate is signed by its own specific CA, so any external connection using
+the IP will get the internal certificate and wont validate against the external
+CA.
+
+If we were to include the node IP in the external certificate we also get an
+immediate match in the mapping, and the hostname verification also works ok.
+All internal connections (those coming from pods through the service network
+IP) will be broken when trying to validate hostname, as the service IP is
+not included. Even if it was included, validation with the internal CA would
+still fail.
+
+Including the IP in the current situation means choosing what to lose:
+internal or external connectivity. Since internal connectivity is mandatory,
+the only option is to use names for external connectivity.
+
+### Merging certificates
+Since having the node IP coexist with trust domain segregation requires not
+including it for external access, there are not too many options left for
+fulfilling the requirements.
+
+Like K3s, which is a competing product, the approach would be to merge all API
+server certificates into a single one, including service names, service IP,
+node IP, node names, and additional SANs from the configuration. The
+certificate matching problem vanishes, as well as the hostname validations and
+different CAs, as it would be a single one.
+
+This approach removes trust domains when accessing API server, like OCP does,
+but the use cases require IP connectivity and authentication from anywhere.
+
+
+### Workflow Description
+Workflow is transparent for the user, as certificates generate themselves upon
+restarting MicroShift.
+
+#### Variation [optional]
+N/A
+
+### API Extensions
+N/A
+
+### Implementation Details/Notes/Constraints [optional]
+N/A
+
+### Risks and Mitigations
+Removal of trust domains and different approach than the one in OpenShift.
+
+Having a single certificate to connect to API server means anyone can
+connect to it, and not having oauth requires extra care when configuring
+cluster roles. This, however, is not too different than the current approach.
+
+Not implementing this change makes MicroShift unable to meet requirements.
+
+### Drawbacks
+N/A
+
+## Design Details
+N/A
+
+### Open Questions [optional]
+N/A
+
+### Test Plan
+N/A
+
+### Graduation Criteria
+N/A
+
+#### Dev Preview -> Tech Preview
+N/A
+
+#### Tech Preview -> GA
+N/A
+
+#### Removing a deprecated feature
+N/A
+
+### Upgrade / Downgrade Strategy
+N/A
+
+### Version Skew Strategy
+N/A
+
+### Operational Aspects of API Extensions
+N/A
+
+#### Failure Modes
+N/A
+
+#### Support Procedures
+N/A
+
+## Implementation History
+N/A
+
+## Alternatives
+* Have several IPs in the node.
+   Having multiple IPs in the node would allow authentication to distinguish
+   between the destination IP and serve the adequate certificate based on what
+   was described before. This would require changing network architecture.
```

Comment:
We have a requirement to be able to deploy MicroShift into environments where we could not ask to have multiple external IPs allocated.

If we were going to change the network architecture, how much would have to change? Can the API server listen on multiple IPs?

Reply:
Right now it listens on any interface: https://github.com/openshift/microshift/blob/main/pkg/controllers/kube-apiserver.go#L181

This alternative would mean having 1 external dedicated IP and one internal for pod-apiserver connections. Then setup internal certificates to use the internal one, and change the kubernetes service endpoint too.

Reply:
Could that internal IP be on one of the bridges visible to the pods but not outside of the host?

Reply:
We will need to be able to run conformance tests with multiple nodes, so it doesn't sound like that solves the whole problem. We could look at using proxies in that test setup, though.